contracts/src/LightClient.sol

```diff
@@ -0,0 +1,152 @@
+        uint64 viewNum;
+        uint64 blockHeight;
+        uint256 blockComm;
+        uint256 feeLedgerComm;
+        uint256 stakeTableBlsKeyComm;
+        uint256 stakeTableSchnorrKeyComm;
+        uint256 stakeTableAmountComm;
+        uint256 threshold;
+    }
+
+    /// @notice Event that a new finalized state has been successfully verified and updated
+    event NewState(uint64 indexed viewNum, uint64 indexed blockHeight, uint256 blockComm);
+
+    /// @notice The state is outdated and older than currently known `finalizedState`
+    error OutdatedState();
+    /// @notice Warning that the last block of the current epoch is not yet submitted before newer
+    /// states are proposed.
+    error MissingLastBlockForCurrentEpoch(uint64 expectedBlockHeight);
+    /// @notice Invalid user inputs: wrong format or non-sensible arguments
+    error InvalidArgs();
+
+    constructor(LightClientState memory genesis, address stakeTableAddr) {
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (newState.viewNum <= finalizedState.viewNum) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // Sanity check (unnecessary under our threat model, just defense in depth)
+        if (newState.blockHeight <= finalizedState.blockHeight) {
+            revert InvalidArgs();
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+
+        // check plonk proof
+        // TODO: (alex) replace the vk with the correct one
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
```

Comment:
When the block is the last one of the epoch, we need to check an extended QC instead of normal QC (see Section 2.4.2 of the HotShot paper). I wonder if we need a special circuit for that case. cc @chancharles92

Reply:
Are we ever going to check QC given the new Schnorr signature flow?

Reply:
> Are we ever going to check QC given the new Schnorr signature flow?

I believe not. The extended QC verification is part of HotShot node software, and honest node won't even sign on the schnorr signature for that special block if unsatisfied.